File: pascal/VMTranslator/CodeWriter.pas
unit CodeWriter;

interface

uses
  Parser, SysUtils;

procedure InitWriter(FileName: string; var f: Text);

procedure WriteInit(var f: Text; n: integer);

procedure WriteArithm(var f: Text; n: integer;
                      ModuleName: string; Operation: string);

procedure WritePush(var f: Text; ModuleName: string;
                                 Segment: string; Index: string);

procedure WritePop(var f: Text; ModuleName: string;
                                Segment: string; Index: string);

procedure WriteGoto(var f: Text; FunctionName: string; LabelName: string);

procedure WriteLabel(var f: Text; FunctionName: string; LabelName: string);

procedure WriteIfGoto(var f: Text; FunctionName: string; LabelName: string);

procedure WriteReturn(var f: Text; Name: string);

procedure WriteCall(var f: Text; n:integer;
                                FunctionName: string; NumArgs: integer);

procedure WriteEnd(var f: Text; Name: string);

procedure WriteRestore(var f: Text; Name: string);

procedure WriteFunction(var f: Text; FunctionName: string; NumVars: integer);

procedure CloseWriter(var f: Text);

implementation

procedure InitWriter(FileName: string; var f: Text);
begin
  assign(f, FileName);
  rewrite(f);
end;

procedure WriteInit(var f: Text; n: integer);
begin
  writeln(f, '// init');
  writeln(f, '@256');
  writeln(f, 'D=A');
  writeln(f, '@SP');
  writeln(f, 'M=D');
  writeln(f);
  WriteCall(f, n, 'Sys.init', 0);
end;

procedure WriteUnary(var f: Text; Operation: string);
begin
  writeln(f, '@SP');
  writeln(f, 'A=M-1');
  if CompareStr(Operation, 'neg') = 0 then
    writeln(f, 'M=-M')
  else
    writeln(f, 'M=!M');
  writeln(f)
end;

procedure WriteRelational(var f: Text; n: integer;
                      ModuleName: string; Operation: string);
begin
  writeln(f, 'D=M-D');
  writeln(f, '@relat.mt.', LowerCase(ModuleName), '.', n);
  writeln(f, 'D;J', UpperCase(Operation));
  writeln(f, '@SP');
  writeln(f, 'A=M-1');
  writeln(f, 'M=0');
  writeln(f, '@relat.end.', LowerCase(ModuleName), '.', n);
  writeln(f, '0;JMP');
  writeln(f, '(relat.mt.', LowerCase(ModuleName), '.', n, ')');
  writeln(f, '@SP');
  writeln(f, 'A=M-1');
  writeln(f, 'M=-1');
  writeln(f, '(relat.end.', LowerCase(ModuleName), '.', n, ')');
end;

procedure WriteBinary(var f: Text; n: integer;
                      ModuleName: string; Operation: string);
begin
  writeln(f, '@SP');
  writeln(f, 'M=M-1');
  writeln(f, 'A=M');
  writeln(f, 'D=M');
  writeln(f, 'A=A-1');
  case Operation of
    'eq', 'gt', 'lt': WriteRelational(f, n, ModuleName, Operation);
    'add': writeln(f, 'M=M+D');
    'sub': writeln(f, 'M=M-D');
    'and': writeln(f, 'M=M&D');
    'or': writeln(f, 'M=M|D');
  end;
  writeln(f);
end;

procedure WriteArithm(var f: Text; n: integer;
                      ModuleName: string; Operation: string);
begin
  writeln(f, '// arithm ', Operation);
  if (CompareStr(Operation, 'neg') = 0)
      or (CompareStr(Operation, 'not') = 0) then
    WriteUnary(f, Operation)
  else
    WriteBinary(f, n, ModuleName, Operation)
end;

function SegmentConstant(Segment: string): string;
begin
  case Segment of
    'local': SegmentConstant := 'LCL';
    'argument': SegmentConstant := 'ARG';
    'this': SegmentConstant := 'THIS';
    'that': SegmentConstant := 'THAT'
  end
end;

procedure WriteVD(var f: Text; ModuleName: string;
                               Segment: string; Index: string);
begin
  case Segment of
    'temp':
    begin
      writeln(f, '@5');
      writeln(f, 'D=A');
      writeln(f, '@', Index);
      writeln(f, 'A=A+D');
      writeln(f, 'D=M');
    end;
    'constant':
    begin
      writeln(f, '@', Index);
      writeln(f, 'D=A');
    end;
    'static':
    begin
      writeln(f, '@', ModuleName, '.', Index);
      writeln(f, 'D=M');
    end;
    'pointer':
    begin
      case Index of
        '0': writeln(f, '@THIS');
        '1': writeln(f, '@THAT');
      end;
      writeln(f, 'D=M');
    end;
    else
    begin
      writeln(f, '@', SegmentConstant(Segment));
      writeln(f, 'D=M');
      writeln(f, '@', Index);
      writeln(f, 'A=A+D');
      writeln(f, 'D=M');
    end;
  end;
end;

procedure WritePushExpr(var f: Text; Expr: string);
begin
  writeln(f, '@SP');
  writeln(f, 'A=M');
  writeln(f, 'M=', Expr);
  writeln(f, '@SP');
  writeln(f, 'M=M+1');
end;

procedure WritePush(var f: Text; ModuleName: string;
                                 Segment: string; Index: string);
begin
  writeln(f, '// push ', Segment, ' ', Index);
  WriteVD(f, ModuleName, Segment, Index);
  WritePushExpr(f, 'D');
  writeln(f)
end;

procedure WriteDV(var f: Text; ModuleName: string;
                               Segment: string; Index: string);
begin
  case Segment of
    'static':
    begin
      writeln(f, '@', ModuleName, '.', Index);
      writeln(f, 'M=D')
    end;
    'pointer':
    begin
      case Index of
        '0': writeln(f, '@THIS');
        '1': writeln(f, '@THAT')
      end;
      writeln(f, 'M=D')
    end
    else
    begin
      writeln(f, '@R13');
      writeln(f, 'M=D');
      if CompareStr(Segment, 'temp') = 0 then
      begin
        writeln(f, '@5');
        writeln(f, 'D=A')
      end
      else
      begin
        writeln(f, '@', SegmentConstant(Segment));
        writeln(f, 'D=M')
      end;
      writeln(f, '@', Index);
      writeln(f, 'D=A+D');
      writeln(f, '@R14');
      writeln(f, 'M=D');
      writeln(f, '@R13');
      writeln(f, 'D=M');
      writeln(f, '@R14');
      writeln(f, 'A=M');
      writeln(f, 'M=D');
      writeln(f)
    end
  end
end;

procedure WritePop(var f: Text; ModuleName: string;
                                Segment: string; Index: string);
begin
  writeln(f, '// pop ', Segment, ' ', Index);
  writeln(f, '@SP');
  writeln(f, 'M=M-1');
  writeln(f, 'A=M');
  writeln(f, 'D=M');
  WriteDV(f, ModuleName, Segment, Index);
  writeln(f);
end;

procedure WriteGoto(var f: Text; FunctionName: string; LabelName: string);
begin
  writeln(f, '// goto ', LabelName);
  writeln(f, '@', FunctionName, '$', LabelName);
  writeln(f, '0;JMP');
  writeln(f);
end;

procedure WriteLabel(var f: Text; FunctionName: string; LabelName: string);
begin
  writeln(f, '// label ', LabelName);
  writeln(f, '(', FunctionName, '$', LabelName, ')');
  writeln(f);
end;

procedure WriteIfGoto(var f: Text; FunctionName: string; LabelName: string);
begin
  writeln(f, '// if-goto ', LabelName);
  writeln(f, '@SP');
  writeln(f, 'M=M-1');
  writeln(f, 'A=M');
  writeln(f, 'D=M');
  writeln(f, '@', FunctionName, '$', LabelName);
  writeln(f, 'D;JNE');
  writeln(f);
end;

procedure WriteReturn(var f: Text; Name: string);
begin
  writeln(f, '// return to ', Name, '$restore');
  writeln(f, '@', Name, '$restore');
  writeln(f, '0;JMP');
  writeln(f);
end;

procedure WriteCall(var f: Text; n:integer; 
                    FunctionName: string; NumArgs: integer);
var
  ReturnLabel: string;
begin
  writeln(f, '// call ', FunctionName, ' ', NumArgs);
  ReturnLabel := FunctionName + '$ret.' + IntToStr(n);

  {push return address}
  writeln(f, '@', ReturnLabel); 
  writeln(f, 'D=A');
  WritePushExpr(f, 'D');

  {push LCL}
  writeln(f, '@LCL');
  writeln(f, 'D=M');
  WritePushExpr(f, 'D');

  {push ARG}
  writeln(f, '@ARG');
  writeln(f, 'D=M');
  WritePushExpr(f, 'D');

  {push THIS}
  writeln(f, '@THIS');
  writeln(f, 'D=M');
  WritePushExpr(f, 'D');

  {push THAT}
  writeln(f, '@THAT');
  writeln(f, 'D=M');
  WritePushExpr(f, 'D');

  {LCL=SP}
  writeln(f, '@SP');
  writeln(f, 'D=M');
  writeln(f, '@LCL');
  writeln(f, 'M=D');

  {ARG=SP-5-NumArgs}
  writeln(f, '@5');
  writeln(f, 'D=D-A');		{D=[LCL]-5}
  writeln(f, '@', NumArgs);
  writeln(f, 'D=D-A');
  writeln(f, '@ARG');
  writeln(f, 'M=D');

  {goto function}
  writeln(f, '@', FunctionName);
  writeln(f, '0;JMP');

  writeln(f, '(', ReturnLabel, ')');

  writeln(f);
end;

procedure WriteEnd(var f: Text; Name: string);
begin
  writeln(f, '(', Name, '$end)');
  writeln(f, '@', Name, '$end');
  writeln(f, '0;JMP');
  writeln(f);
end;

procedure WriteRestore(var f: Text; Name: string);
begin
  writeln(f, '(', Name, '$restore)');

  writeln(f, '@ARG');		{save ARG}
  writeln(f, 'D=M');
  writeln(f, '@R14');
  writeln(f, 'M=D');

  writeln(f, '@LCL');
  writeln(f, 'D=M-1');  {D=[LCL]-1}
  writeln(f, '@R13');
  writeln(f, 'AM=D');	  {[R13]=[LCL]-1}
  writeln(f, 'D=M');    {D=[[LCL]-1]}
  writeln(f, '@THAT');
  writeln(f, 'M=D');

  writeln(f, '@R13');
  writeln(f, 'D=M-1');  {D=[LCL]-2}
  writeln(f, 'AM=D');		{[R13]=[LCL]-2}
  writeln(f, 'D=M');    {D=[[LCL]-2]}
  writeln(f, '@THIS');
  writeln(f, 'M=D');

  writeln(f, '@R13');
  writeln(f, 'D=M-1');  {D=[LCL]-3}
  writeln(f, 'AM=D');		{[R13]=[LCL]-3}
  writeln(f, 'D=M');    {D=[[LCL]-3]}
  writeln(f, '@ARG');
  writeln(f, 'M=D');

  writeln(f, '@R13');
  writeln(f, 'D=M-1');  {D=[LCL]-4}
  writeln(f, 'AM=D');		{[R13]=[LCL]-4}
  writeln(f, 'D=M');    {D=[[LCL]-4]}
  writeln(f, '@LCL');
  writeln(f, 'M=D');

  writeln(f, '@R13');		{save return address}
  writeln(f, 'A=M-1');
  writeln(f, 'D=M');		{D=[[LCL]-5]}
  writeln(f, '@R13');
  writeln(f, 'M=D');		{[R13]=[[LCL]-15]}

  writeln(f, '@SP');		{set return value}
  writeln(f, 'A=M-1');
  writeln(f, 'D=M');
  writeln(f, '@R14');
  writeln(f, 'A=M');
  writeln(f, 'M=D');		{[[old-ARG]]=[[SP]-1]}

  writeln(f, '@R14');   {set [SP] after return value}
  writeln(f, 'D=M+1');
  writeln(f, '@SP');
  writeln(f, 'M=D');    {[SP]=[old-ARG]+1}

  writeln(f, '@R13');
  writeln(f, 'A=M');
  writeln(f, '0;JMP');
  writeln(f);
end;


procedure WriteFunction(var f: Text; 
                                FunctionName: string; NumVars: integer);
var
  i: integer;
begin
  writeln(f, '// function ', FunctionName, ' ', NumVars);
  writeln(f, '(', FunctionName, ')');

  for i := 1 to NumVars do
    WritePushExpr(f, '0');

  writeln(f);
end;

procedure CloseWriter(var f: Text);
begin
  close(f);
end;

end.

File: java/VMTranslator/CodeWriter.java
import java.io.*;

public class CodeWriter {

    private String name;
    private PrintStream printStream;
    private String moduleName;
    private String functionName;
    private int n;

    public CodeWriter(File file) {
        try {
            printStream = new PrintStream(file);
        } catch (IOException e) {
            e.printStackTrace();
        }
        n = 0;
        functionName = "Sys.init";
        name = VMTranslator.getNoExtName(file);
    }

    public void incN() {
        n++;
    }

    public void setModuleName(String moduleName) {
        this.moduleName = moduleName;
    }

    public void writeInit() {
        printStream.println("// init");
        printStream.println("@256");
        printStream.println("D=A");
        printStream.println("@SP");
        printStream.println("M=D");
        printStream.println();
        writeCall("Sys.init", 0);
    }

    private void writeUnary(String operation) {
        printStream.println("@SP");
        printStream.println("A=M-1");
        if (operation.equals("neg")) {
            printStream.println("M=-M");
        } else {
            printStream.println("M=!M");
        }
        printStream.println();
    }

    private void writeRelational(String operation) {
        printStream.println("D=M-D");
        printStream.println("@relat.mt." + moduleName + "." + n);
        printStream.println("D;J" + operation.toUpperCase());
        printStream.println("@SP");
        printStream.println("A=M-1");
        printStream.println("M=0");
        printStream.println("@relat.end." + moduleName + "." + n);
        printStream.println("0;JMP");
        printStream.println("(relat.mt." + moduleName + "." + n + ")");
        printStream.println("@SP");
        printStream.println("A=M-1");
        printStream.println("M=-1");
        printStream.println("(relat.end." + moduleName + "." + n + ")");
    }


    private void writeBinary(String operation) {
        printStream.println("@SP");
        printStream.println("M=M-1");
        printStream.println("A=M");
        printStream.println("D=M");
        printStream.println("A=A-1");
        switch (operation) {
            case "eq":
            case "gt":
            case "lt":
                writeRelational(operation);
                break;
            case "add":
                printStream.println("M=M+D");
                break;
            case "sub":
                printStream.println("M=M-D");
                break;
            case "and":
                printStream.println("M=M&D");
                break;
            case "or":
                printStream.println("M=M|D");
                break;
            default:
                break;
        }
        printStream.println();
    }

    public void writeArithm(String operation) {
        printStream.println("// arithm " + operation);
        if (operation.equals("neg") || operation.equals("not")) {
            writeUnary(operation);
        } else {
            writeBinary(operation);
        }
    }

    private String segmentConstant(String segment) {
        switch (segment) {
            case "local":
                return "LCL";
            case "argument":
                return "ARG";
            case "this":
                return "THIS";
            case "that":
                return "THAT";
            default:
                return null;
        }
    }

    private void writeVD(String segment, int index) {
        switch (segment) {
            case "temp":
                printStream.println("@5");
                printStream.println("D=A");
                printStream.println("@" + index);
                printStream.println("A=A+D");
                printStream.println("D=M");
                break;
            case "constant":
                printStream.println("@" + index);
                printStream.println("D=A");
                break;
            case "static":
                printStream.println("@" + moduleName + "." + index);
                printStream.println("D=M");
                break;
            case "pointer":
                if (index == 0) {
                    printStream.println("@THIS");
                } else {
                    printStream.println("@THAT");
                }
                printStream.println("D=M");
                break;
            default:
                printStream.println("@" + segmentConstant(segment));
                printStream.println("D=M");
                printStream.println("@" + index);
                printStream.println("A=A+D");
                printStream.println("D=M");
                break;
        }
    }

    private void writePushExpr(String expr) {
        printStream.println("@SP");
        printStream.println("A=M");
        printStream.println("M=" + expr);
        printStream.println("@SP");
        printStream.println("M=M+1");
    }

    public void writePush(String segment, int index) {
        printStream.println("// push " + segment + " " + index);
        writeVD(segment, index);
        writePushExpr("D");
        printStream.println();
    }

    private void writeDV(String segment, int index) {
        switch (segment) {
            case "static":
                printStream.println("@" + moduleName + "." + index);
                printStream.println("M=D");
                break;
            case "pointer":
                if (index == 0) {
                    printStream.println("@THIS");
                } else {
                    printStream.println("@THAT");
                }
                printStream.println("M=D");
                break;
            default:
                printStream.println("@R13");
                printStream.println("M=D");
                if (segment.equals("temp")) {
                    printStream.println("@5");
                    printStream.println("D=A");
                } else {
                    printStream.println("@" + segmentConstant(segment));
                    printStream.println("D=M");
                }
                printStream.println("@" + index);
                printStream.println("D=A+D");
                printStream.println("@R14");
                printStream.println("M=D");
                printStream.println("@R13");
                printStream.println("D=M");
                printStream.println("@R14");
                printStream.println("A=M");
                printStream.println("M=D");
                printStream.println();
                break;
        }
    }

    public void writePop(String segment, int index) {
        printStream.println("// pop " + segment + " " + index);
        printStream.println("@SP");
        printStream.println("M=M-1");
        printStream.println("A=M");
        printStream.println("D=M");
        writeDV(segment, index);
        printStream.println();
    }

    public void writeGoto(String labelName) {
        printStream.println("// goto " + labelName);
        printStream.println("@" + functionName + "$" + labelName);
        printStream.println("0;JMP");
        printStream.println();
    }

    public void writeLabel(String labelName) {
        printStream.println("// label " + labelName);
        printStream.println("(" + functionName + "$" + labelName + ")");
        printStream.println();
    }

    public void writeIfGoto(String labelName) {
        printStream.println("// if-goto " + labelName);
        printStream.println("@SP");
        printStream.println("M=M-1");
        printStream.println("A=M");
        printStream.println("D=M");
        printStream.println("@" + functionName + "$" + labelName);
        printStream.println("D;JNE");
        printStream.println();
    }

    public void writeReturn() {
        printStream.println("// return to " + name + "$restore");
        printStream.println("@" + name + "$restore");
        printStream.println("0;JMP");
        printStream.println();
    }

    public void writeCall(String func, int nVars) {
        printStream.println("// call " + func + " " + nVars);
        String returnLabel = func + "$ret." + n;

        /* push return address */
        printStream.println("@" + returnLabel);
        printStream.println("D=A");
        writePushExpr("D");

        /* push LCL */
        printStream.println("@LCL");
        printStream.println("D=M");
        writePushExpr("D");

        /* push ARG */
        printStream.println("@ARG");
        printStream.println("D=M");
        writePushExpr("D");

        /* push THIS */
        printStream.println("@THIS");
        printStream.println("D=M");
        writePushExpr("D");

        /* push THAT */
        printStream.println("@THAT");
        printStream.println("D=M");
        writePushExpr("D");

        /* LCL=SP */
        printStream.println("@SP");
        printStream.println("D=M");
        printStream.println("@LCL");
        printStream.println("M=D");

        /* ARG=SP-5-NumArgs */
        printStream.println("@5");
        printStream.println("D=D-A");        /* D=[LCL]-5 */
        printStream.println("@" + nVars);
        printStream.println("D=D-A");
        printStream.println("@ARG");
        printStream.println("M=D");

        /* goto function */
        printStream.println("@" + func);
        printStream.println("0;JMP");

        printStream.println("(" + returnLabel + ")");

        printStream.println();
    }

    public void writeEndLoop() {
        printStream.println("(" + name + "$end)");
        printStream.println("@" + name + "$end");
        printStream.println("0;JMP");
        printStream.println();
    }

    public void writeRestore() {
        printStream.println("(" + name + "$restore)");

        printStream.println("@ARG");    /* save ARG */
        printStream.println("D=M");
        printStream.println("@R14");
        printStream.println("M=D");

        printStream.println("@LCL");
        printStream.println("D=M-1");  /* D=[LCL]-1 */
        printStream.println("@R13");
        printStream.println("AM=D");   /* [R13]=[LCL]-1 */
        printStream.println("D=M");    /* D=[[LCL]-1] */
        printStream.println("@THAT");
        printStream.println("M=D");

        printStream.println("@R13");
        printStream.println("D=M-1");  /* D=[LCL]-2 */
        printStream.println("AM=D");        /* [R13]=[LCL]-2 */
        printStream.println("D=M");    /* D=[[LCL]-2] */
        printStream.println("@THIS");
        printStream.println("M=D");

        printStream.println("@R13");
        printStream.println("D=M-1");  /* D=[LCL]-3 */
        printStream.println("AM=D");   /* [R13]=[LCL]-3 */
        printStream.println("D=M");    /* D=[[LCL]-3] */
        printStream.println("@ARG");
        printStream.println("M=D");

        printStream.println("@R13");
        printStream.println("D=M-1");  /* D=[LCL]-4 */
        printStream.println("AM=D");   /* [R13]=[LCL]-4 */
        printStream.println("D=M");    /* D=[[LCL]-4] */
        printStream.println("@LCL");
        printStream.println("M=D");

        printStream.println("@R13");    /* save return address */
        printStream.println("A=M-1");
        printStream.println("D=M");     /* D=[[LCL]-5] */
        printStream.println("@R13");
        printStream.println("M=D");     /* [R13]=[[LCL]-15] */

        printStream.println("@SP");     /* set return value */
        printStream.println("A=M-1");
        printStream.println("D=M");
        printStream.println("@R14");
        printStream.println("A=M");
        printStream.println("M=D");     /* [[old-ARG]]=[[SP]-1] */

        printStream.println("@R14");    /* set [SP] after return value */
        printStream.println("D=M+1");
        printStream.println("@SP");
        printStream.println("M=D");    /* [SP]=[old-ARG]+1 */

        printStream.println("@R13");
        printStream.println("A=M");
        printStream.println("0;JMP");
        printStream.println();
    }

    public void writeFunction(String func, int nArgs) {
        functionName = func;
        printStream.println("// function " + functionName + " " + nArgs);
        printStream.println("(" + functionName + ")");

        for (int i = 0; i < nArgs; i++) {
            writePushExpr("0");
        }

        printStream.println();
    }

    public void close() {
        printStream.close();
    }
}
